### ISSUE TITLE
Stack overflow while decompiling 16-bit DOS executable

### ISSUE THREAD

**Issue by erenes** [uxmal/reko#850]
I am trying to decompile the GOG version of Wacky Wheels (https://www.gog.com/game/wacky_wheels). 

While this does appear to give valid assembler code, when I move to the "Analyze dataflow" step, Reko crashes. 

After setting up my debug environment I found that I ran into a Stack overflow when reading the following instruction: 
![image](https://user-images.githubusercontent.com/3177915/78711648-9c32b900-7917-11ea-864b-6572c7e9a0f8.png)
relevant part of the asm: [ww.asm.txt](https://github.com/uxmal/reko/files/4446750/ww.asm.txt)
relevant part of the dis: [ww.dis.txt](https://github.com/uxmal/reko/files/4446751/ww.dis.txt)

The stack trace of the exception: 
[stacktrace.txt](https://github.com/uxmal/reko/files/4446678/stacktrace.txt)

My assembler knowledge isn't great, but it definitely looks to me an infinite loop has been generated in the asm code. Perhaps it's an upstream issue in Capstone.

Unfortunately all unit tests are failing in my environment (due to line-ending issues, probably caused by my git setup) so it's currently not possible for me to write a unit test for this on my own. 
I have "fixed" the crash in my fork by just counting the number of recursions and disallowing anything above an (arbitrary) 25 levels of recursion: 
https://github.com/erenes/reko/commit/a77067b348bbd8ca139aa617f5989f901e703eaa

Please let me know if there is any information I can provide to further help investigate.

**Comment by uxmal** [comment]
Thanks for reporting this. I will take a look tomorrow to see if I can find the cause.

**Comment by uxmal** [comment]
One thing you could to assist is to enable verbose tracing when the `SsaTransformer` runs on the procedure with the problem. For instance, in the constructor of `SsaTransformer` in `src/Decompiler/Analysis/SsaTransformer.cs`, you could add the following line:
```C#
            trace.Level = proc.Name == "fnXXXX_XXXX" ? TraceLevel.Verbose : TraceLevel.Off;
```
where `fnXXXX_XXXX` is the name of the procedure that's exhibiting the bad behavior. This will generate debug output in the `Output` window of Visual Studio. This will show the basic blocks where the algorithm is spinning, along with which identifier it's failing to resolve.

**Comment by erenes** [comment]
The following is all trace output from the moment of hitting the constructor of SsaTransformer (and setting loglevel to verbose) until the StackOverflowException (I still have the 25 levels of recursion patch enabled to avoid a log that's a bit TOO verbose):
[verboseTrace.txt](https://github.com/uxmal/reko/files/4451239/verboseTrace.txt)

**Comment by uxmal** [comment]
Could you add a 
```C#
trace.Warning("*** Exceeded 25 levels of recursion, stopping");
```
at the place where you quit the recursion and run this again? I'd like to see the output right before that message would be output.

**Comment by erenes** [comment]
I actually put a breakpoint there and copied all output that's before that point. 
Nevertheless, this is another version without breakpoints and including a warning. 

FYI, I cleared the debug window before clicking "analyze dataflow": 
[verboseTrace2.txt](https://github.com/uxmal/reko/files/4451374/verboseTrace2.txt)

**Comment by uxmal** [comment]
Something is very wrong with the exit block of that procedure. Is there any way you could provide me with the binary in question? Contact me on https://gitter.im/uxmal/reko to  "close the loop"

**Comment by uxmal** [comment]
I'm closing this issue now, as together with the dcproj I send you in gitter.im, you should be able to make progress. The stack overflow will happen any time the output of the scanner phase returns broken control flow graphs (CFGs). I fixed a few issued caused by missing rewriters/lifters of x86 instructions (notably `lldt` and `lmsw`). 

However, there is still a remaining big issue, the scanner is the oldest part of Reko and is showing its  age. Specifically, its eager approach to building `Procedures` is hard to follow and buggy. I'm currently rewriting parts of the scanner, which will hopefully make most of these problems go away.

Thanks again for reporting this issue!
